src/ask-password/ask-password.c

```diff
@@ -232,9 +240,163 @@ static int parse_argv(int argc, char *argv[]) {
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "message",      SD_JSON_VARIANT_STRING,        sd_json_dispatch_const_string, offsetof(MethodAskParameters, message),       0 },
+                { "keyname",      SD_JSON_VARIANT_STRING,        sd_json_dispatch_const_string, offsetof(MethodAskParameters, keyring),       0 },
+                { "icon",         SD_JSON_VARIANT_STRING,        sd_json_dispatch_const_string, offsetof(MethodAskParameters, icon),          0 },
+                { "id",           SD_JSON_VARIANT_STRING,        sd_json_dispatch_const_string, offsetof(MethodAskParameters, id),            0 },
+                { "timeoutUSec",  _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64,       offsetof(MethodAskParameters, timeout_usec),  0 },
+                { "untilUSec",    _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64,       offsetof(MethodAskParameters, until_usec),    0 },
+                { "acceptCached", SD_JSON_VARIANT_BOOLEAN,       sd_json_dispatch_tristate,     offsetof(MethodAskParameters, accept_cached), 0 },
+                { "pushCache",    SD_JSON_VARIANT_BOOLEAN,       sd_json_dispatch_tristate,     offsetof(MethodAskParameters, push_cache) ,   0 },
+                { "echo",         SD_JSON_VARIANT_STRING,        dispatch_echo_mode,            offsetof(MethodAskParameters, echo_mode),     0 },
+                VARLINK_DISPATCH_POLKIT_FIELD,
+                {}
+        };
+
+        Hashmap **polkit_registry = ASSERT_PTR(userdata);
+        MethodAskParameters p = {
+                .timeout_usec = DEFAULT_TIMEOUT_USEC,
+                .until_usec = UINT64_MAX,
+                .accept_cached = -1,
+                .push_cache = -1,
+                .echo_mode = _ECHO_MODE_INVALID,
+        };
+        int r;
+
+        assert(link);
+
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
+        if (r != 0)
+                return r;
+
+        r = varlink_verify_polkit_async_full(
+                        link,
+                        /* bus= */ NULL,
+                        "io.systemd.ask-password.ask",
+                        /* details= */ NULL,
+                        /* good_user= */ FLAGS_SET(arg_flags, ASK_PASSWORD_USER) ? getuid() : UID_INVALID,
+                        /* flags= */ 0,
+                        polkit_registry);
+        if (r <= 0)
+                return r;
+
+        AskPasswordRequest req = {
+                .tty_fd = -EBADF,
+                .message = p.message ?: arg_message,
+                .icon = p.icon ?: arg_icon,
+                .id = p.id ?: arg_id,
+                .keyring = p.keyring ?: arg_key_name,
+                .credential = arg_credential_name,
+                .hup_fd = sd_varlink_get_input_fd(link),
+        };
+
+        if (p.timeout_usec != 0 && p.until_usec != 0)
+                req.until = MIN(usec_add(now(CLOCK_MONOTONIC), p.timeout_usec), p.until_usec);
```

Comment:
`p.timeout_usec` is only used here. Why two similar timeout parameters are introduced??

Reply:
one is relative the other is absolute. 

depending on the usecase one or the other might be better. i generally think passing absolute times is generally better, but then again, for remotability relative times are much better (since a remote app might not know CLOCK_MONOTONIC on the target machine). Hence I figured we should just support both. The line above makes sure sure to pick the closer timeout and propagate it into the actual request.